java/ql/src/Security/CWE/CWE-327/MaybeBrokenCryptoAlgorithm.ql

```diff
@@ -27,7 +27,9 @@ import InsecureCryptoFlow::PathGraph
 string getStringValue(DataFlow::Node algo) {
   result = algo.asExpr().(StringLiteral).getValue()
   or
-  result = algo.asExpr().(PropertiesGetPropertyMethodCall).getPropertyValue()
+  exists(string value | value = algo.asExpr().(PropertiesGetPropertyMethodCall).getPropertyValue() |
+    result = value and not value.regexpMatch(getSecureAlgorithmRegex())
+  )
 }
```

Comment:
Having to keep this manually synchronized with `InsecureCryptoConfig::isSource` seems like a potential footgun.

What about adding this class:

```codeql
/** A reference to an insecure cryptographic algorithm. */
abstract class InsecureAlgorithm extends Expr {
  /** Gets the string representation of this insecure cryptographic algorithm. */
  abstract string getStringValue();
}
```

That both `InsecureAlgoLiteral` and a new `InsecureAlgoProperty` can extend:

```codeql
class InsecureAlgoLiteral extends InsecureAlgorithm, ShortStringLiteral {
  // ...

  override string getStringValue() { result = this.getValue() }
}

/**
 * A property access that may refer to an insecure cryptographic algorithm.
 */
class InsecureAlgoProperty extends InsecureAlgorithm, PropertiesGetPropertyMethodCall {
  string value;

  InsecureAlgoProperty() {
    value = this.getPropertyValue() and
    // Since properties pairs are not included in the java/weak-cryptographic-algorithm,
    // the check for values from properties files can be less strict than `InsecureAlgoLiteral`.
    not value.regexpMatch(getSecureAlgorithmRegex())
  }

  override string getStringValue() { result = value }
}
```

And that can be used in `InsecureCryptoConfig::isSource`:

```codeql
predicate isSource(DataFlow::Node n) { n.asExpr() instanceof InsecureAlgorithm }
```

That way, we could get rid of `getStringValue` completely and just do:

```codeql
from InsecureCryptoFlow::PathNode source, InsecureCryptoFlow::PathNode sink, CryptoAlgoSpec c
where
  sink.getNode().asExpr() = c.getAlgoSpec() and
  InsecureCryptoFlow::flowPath(source, sink)
select c, source, sink,
  "Cryptographic algorithm $@ may not be secure, consider using a different algorithm.", source,
  source.getNode().asExpr().(InsecureAlgorithm).getStringValue()
```

If this sounds right to you, I've created a PR that you can merge into your branch to apply these changes: https://github.com/egregius313/codeql/pull/99.

Reply:
DCA looks good on that PR.